docs/indexing.md

```diff
@@ -0,0 +1,422 @@
+
+## **Function and Domain**
+
+The indexing map is a function $\boldsymbol{f}(\boldsymbol{d}, \boldsymbol{s})$ that maps a multi-index  $\boldsymbol{d}$ of a tensor $A$ to elements/ranges of tensor $B$. The parameter $\boldsymbol{s}$ refers to the ranges of indices of  the dimensions that are present in tensor $B$, but not in tensor $A$​.
+
+For example, if we have a reduction from `tensor<2x4x8x16xf32>` to `tensor<4x8xf32>`, then the indexing map from the 2D output to the 4D input is $(d_0, d_1) \mapsto (s_0, d_0, d_1, s_1)$, where $d_i$ are the dimension parameters that correspond to the indices of the output tensor. Parameters $s_j$ encode multiple values, i.e. to compute a $(d_0, d_1)$ element of the output, we need $(s_0, d_0, d_1, s_1)$ elements of the input, where $s_0 \in [0, 2)$ and $s_1 \in [0, 16)$. 
+
+This mapping can be constructed from the attributes of HLO instructions or the mappings of unfused instructions can be composed to get indexing for a fusion. The mapping also has a domain, which specifies for what elements of the tensor the mapping exists. 
+$$
+\begin{eqnarray}
+\boldsymbol{f}(\boldsymbol{d}, \boldsymbol{s})\; &s.t.& \\
+ \boldsymbol{lb}_d &\leq& \boldsymbol{d} \leq \boldsymbol{ub}_d \\
+  \boldsymbol{lb}_s &\leq& \boldsymbol{s} \leq \boldsymbol{ub}_s \\
+   \boldsymbol{lb}_g &\leq& \boldsymbol{g}(\boldsymbol{d}, \boldsymbol{s}) \leq \boldsymbol{ub}_g
+\end{eqnarray}
+$$
+Since we want to minimize recomputation, we need a library for symbolic computations. XLA already depends on MLIR, so we use [mlir::AffineMap](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/AffineMap.h) instead of writing a symbolic arithmetic library.
+
+A typical `AffineMap` looks like 
+
+```
+(d0)[s0, s1] -> (s0 + 5, d0 * 2, s1 * 3 + 50)
+```
+
+`AffineMap` conveniently has two types of parameters: *dimensions* and *symbols* that we can use for $\boldsymbol d$ and $\boldsymbol s$ respectively.  `AffineMap` does not contain any metadata about ranges of the dimensions, so we have to provide this data ourselves.
+
+```c++
+struct Range {
+ int64_t lower_bound;
+ int64_t upper_bound;
+};
+
+struct Domain {
+ std::vector<Range> dimension_ranges;
+ std::vector<Range> symbol_ranges;
+ llvm::DenseMap<mlir::AffineExpr, Range> expr_ranges;
+};
+
+struct IndexingMap {
+ mlir::AffineMap affine_map;
+ Domain domain;
+};
+
+```
+
+`dim_ranges` encodes the **inclusive** box constraints for the dimension parameters $\boldsymbol{d}$ of the indexing map, which usually coincide with the shape of the output tensor for ops like transpose-reduce-elementwise-dot, but there are some exceptions like  [HloConcatenateInstruction](https://github.com/openxla/xla/blob/578b6df240be94c3c84129fd83f34487efc623a5/xla/hlo/ir/hlo_instructions.h#L927).
```

Comment:
Nit: " transpose-reduce-elementwise-dot" - commas rather than hyphens?

The operational semantics of concatenate are spec'ed here:
https://github.com/openxla/stablehlo/blob/main/docs/spec.md#concatenate
Maybe more useful than the C++ class definition?